rails_programming/turbo/stimulus_js.md

```diff
@@ -0,0 +1,287 @@
+a specific CSS class on an element. But the class might be different every time, for example you might sometimes want
+to toggle a `hidden` class, but sometimes an `active` one. For these situations you can configure your controller
+with an attribute to specific the CSS class to be used:
+
+````html
+<div data-controller="toggle" data-toggle-change-class="hidden"></div>
+````
+
+And in your controller:
+```js
+// app/javascript/controllers/toggle_controller.js
+export default class extends Controller {
+  static classes = [ "change" ]
+
+  toggle() {
+    this.element.classList.toggle(this.changeClass)
+  }
+}
+
+```
+
+### Set up stuff with the lifecycle functions
+
+Often you will want to use Stimulus to use third party javascript libraries with your code. For example to turn your
+normal select form input into a fancy interactive field with autocompletion and so forth. This means you want to execute
+some JavaScript whenever you got such a field.
+
+````html
+<select name="foods" id="foods" data-controller="select">
+  <option value="Pizza">Pizza</option>
+  ...
+</select>
+````
+
+In this situation you don't have a user do anything you could react to, he probably just came from another page. For
+these situation you can use the lifecycle methods Stimulus. We'll cover only the most important one for now: `connect`
+
+````javascript
+import { Controller } from "@hotwired/stimulus"
+import Choices from 'choices.js'
+
+export default class extends Controller {
+  connect() {
+    new Choices(this.element)
+  }
+}
+````
+So connect is a special function, that gets called whenever an element with the `data-controller="controller-name"`
+appears in the DOM. So the perfect solution to change things about the html that just appears on the page, that you
+need to change somehow.
+
+### Summary
+
+Stimulus gives you a way to make your html more interactive, by using data attributes your html makes visible where
+your Javascript will interact with your HTML. This was just a broad introduction, please follow the assignments that will
+go more into depth.
+
+### Assignment
+
+<div class="lesson-content__panel" markdown="1">
+* Read the [Stimulus Handbook](https://stimulus.hotwired.dev/handbook/origin) to get an overview of how to code
+with Stimulus
+* Watch this [Stimulus 2.0 Tutorial Video](https://www.driftingruby.com/episodes/the-stimulus-2-0-tutorial) it may give
+you a bit of a feel on how to work with Stimulus controllers. You can ignore the part about installation with Webpacker
+we will use the new rails standard of using import maps.
+* Make sure to also read the [reference section](https://stimulus.hotwired.dev/reference/lifecycle-callbacks), if you 
+haven't already, don't worry if not everything sticks, but you should know where to look up these things.
+</div>
+
+### Exercises
+
+To practice you create a new standard rails application or play around on a separate branch on one of your existing
```

Comment:
We probably just want them to spin up a new application. If they butcher the branch and accidentally merge it we'd have all kinds of support questions to answer 🙈

Reply:
Good point!